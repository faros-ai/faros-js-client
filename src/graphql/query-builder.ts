import {EnumType, jsonToGraphQLQuery} from 'json-to-graphql-query';
import {isNil} from 'lodash';

import {
  ConflictClause,
  DeleteMutationObject,
  Mutation,
  UpsertMutationObject,
} from './types';

type MutationFieldValue =
  | string
  | number
  | boolean
  | any[]
  | {category: string; detail: string}
  | {[field: string]: string | number | boolean}
  | Ref
  | undefined
  | null;

interface MutationFields {
  [field: string]: MutationFieldValue;
}

interface Conflict {
  constraint: string;
  update_columns: string[];
}

export interface FarosModel {
  [modelName: string]: MutationFields;
}

export class Ref {
  constructor(readonly model?: FarosModel) {}
}

export class QueryBuilder {
  constructor(private readonly origin: string) {}

  /**
   * Creates an upsert mutation for the provided Faros model.
   * @param model   The Faros model
   * @param conflict Override the default conflict clause
   * @returns       The upsert mutation
   */
  upsert(model: FarosModel, conflict?: Conflict): Mutation {
    const mutationObj = this.upsertMutationObj({model, ref: false, conflict});
    const modelName = Object.keys(model)[0];
    return {
      mutation: {
        [`insert_${modelName}_one`]: {__args: mutationObj, id: true},
      },
    };
  }

  delete(model: FarosModel): Mutation {
    const deleteObj = this.deleteMutationObj(model);
    const modelName = Object.keys(model)[0];
    return {
      mutation: {
        [`delete_${modelName}`]: {
          __args: deleteObj,
          returning: {
            id: true,
          },
        },
      },
    };
  }

  /**
   * Creates a Ref that can be used in another Faros Model.
   */
  ref(model?: FarosModel): Ref {
    return new Ref(model);
  }

  /**
   * Creates a mutation object that will update every field of the Faros Model.
   * If the model contains fields that reference another
   * model, those fields will be recursively turned into MutationReferences.
   * @param model   The Faros model
   * @param ref     If the mutationObj should be a reference
   * @returns       The mutation object
   */
  private upsertMutationObj(args: {
    model: FarosModel;
    conflict?: Conflict;
    ref?: boolean;
  }): UpsertMutationObject {
    const {model, conflict} = args;
    const ref = args.ref ?? false;

    const [modelName, fields] = Object.entries(model)[0];
    const mutObj: any = {};
    const mask = ['refreshedAt'];

    for (const [k, v] of Object.entries(fields ?? {})) {
      let maskKey = k;
      if (isNil(v)) {
        mutObj[k] = null;
      } else if (v instanceof Ref) {
        if (v.model) {
          mutObj[k] = this.upsertMutationObj({model: v.model, ref: true});
        } else {
          mutObj[k] = null;
        }
        // ref's key should be suffixed with Id for onConflict field
        maskKey += 'Id';
      } else {
        mutObj[k] = Array.isArray(v) ? arrayLiteral(v) : v;
      }
      if (!ref) {
        mask.push(maskKey);
      }
    }

    if (!ref) {
      mutObj.origin = this.origin;
      conflict?.update_columns?.push('origin');
      mask.push('origin');
    }

    const defaultConflict = {
      constraint: `${modelName}_pkey`,
      update_columns: mask,
    };

    return {
      [!ref ? 'object' : 'data']: mutObj,
      on_conflict: this.createConflictClause(conflict ?? defaultConflict),
    };
  }

  /**
   * Creates a delete mutation object that will create _eq keys for every field
   * for the Faros Model.
   *
   * @param model   The Faros model
   * @param ref     If the DeleteMutationObject is a reference
   * @returns       The mutation object
   */
  private deleteMutationObj(
    model: FarosModel,
    ref = false
  ): DeleteMutationObject {
    const fields = Object.values(model)[0];
    const mutObj: any = {};

    for (const [k, v] of Object.entries(fields ?? {})) {
      if (isNil(v)) {
        mutObj[k] = null;
      } else if (v instanceof Ref) {
        if (v.model) {
          mutObj[k] = this.deleteMutationObj(v.model, true);
        } else {
          mutObj[k + 'Id'] = {_is_null: true};
        }
      } else {
        const value = Array.isArray(v) ? arrayLiteral(v) : v;
        mutObj[k] = {_eq: value};
      }
    }

    return !ref ? {where: mutObj} : mutObj;
  }

  private createConflictClause(conflict: Conflict): ConflictClause {
    return {
      constraint: new EnumType(conflict.constraint),
      update_columns: conflict.update_columns.map((c) => new EnumType(c)),
    };
  }
}

/**
 * Convert string arrays into postgres literal array.
 * Leave non-string arrays untouched.
 */
export function arrayLiteral(arr: any[]): string | object {
  for (const item of arr) {
    if (typeof item !== 'string') {
      return arr;
    }
  }
  return `{${arr
    .map((s) => {
      const pre = s.startsWith('"') ? '' : '"';
      const suf = s.endsWith('"') ? '' : '"';
      return `${pre}${s}${suf}`;
    })
    .join(',')}}`;
}

export function mask(object: any): string[] {
  return Object.keys(object);
}

/**
 * Constructs a gql query from an array of json mutations.
 * The outputted qql mutation might look like:
 *
 *   mutation  {
 *     i1: insert_cicd_Artifact_one(object: {uid: "u1b"}) {
 *       id
 *       refreshedAt
 *     }
 *     i2: insert_cicd_Artifact_one(object: {uid: "u2b"}) {
 *       id
 *       refreshedAt
 *     }
 *   }
 *
 *  Notable here are the i1/i2 aliases.
 *  These are required when multiple operations share the same
 *  name (e.g. insert_cicd_Artifact_one) and are supported in
 *  jsonToGraphQLQuery with __aliasFor directive.
 *
 *  @return batch gql mutation or undefined if the input is undefined, empty
 *  or doesn't contain any mutations.
 */
export function batchMutation(mutations: Mutation[]): string | undefined {
  if (mutations.length) {
    const queryObj: any = {};
    mutations.forEach((query, idx) => {
      if (query.mutation) {
        const queryType = Object.keys(query.mutation)[0];
        const queryBody = query.mutation[queryType];
        queryObj[`m${idx}`] = {
          __aliasFor: queryType,
          ...queryBody,
        };
      }
    });
    if (Object.keys(queryObj).length > 0) {
      return jsonToGraphQLQuery({mutation: queryObj});
    }
  }
  return undefined;
}
