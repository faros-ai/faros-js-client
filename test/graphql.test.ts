import 'jest-extended';

import * as gql from 'graphql';

import * as sut from '../src/graphql/graphql';
import {
  graphSchemaV2,
  graphSchemaV2ForForeignKeyExclusionTest,
  graphSchemaV2ForPrimaryKeysTest,
  loadQueryFile,
  toArray,
} from './helpers';

describe('graphql', () => {
  test('flatten nodes V2', async () => {
    const nodes = [
      {
        id: 'fc419273d9256727b0970326769edf1b99ae55da',
        author: {
          ownerId: 't1|ted|GitHub|nat-casey',
        },
      },
      {
        id: '32a4d92f46f02e9c10cca58e55874b6269690891',
        author: {
          ownerId: 't1|ted|GitHub|ypc-faros',
        },
      },
    ];
    const query = await loadQueryFile('commits-v2.gql');
    const ctx = sut.flattenV2(query, graphSchemaV2);
    const flattenedNodes = sut.flattenIterable(ctx, nodes);
    expect(ctx.fieldTypes).toEqual(
      new Map([
        ['id', gql.GraphQLString],
        ['owner_id', gql.GraphQLString],
      ])
    );
    expect(Array.from(ctx.params.keys())).toIncludeSameMembers(['from', 'to']);
    function assertTypeAndValue(fld: string, type: string) {
      expect(gql.isScalarType(ctx.params.get(fld))).toBeTrue();
      expect((ctx.params.get(fld) as gql.GraphQLScalarType).name).toEqual(type);
    }
    assertTypeAndValue('to', 'timestamptz');
    assertTypeAndValue('from', 'timestamptz');
    expect(await toArray(flattenedNodes)).toIncludeSameMembers([
      {
        id: 'fc419273d9256727b0970326769edf1b99ae55da',
        owner_id: 't1|ted|GitHub|nat-casey',
      },
      {
        id: '32a4d92f46f02e9c10cca58e55874b6269690891',
        owner_id: 't1|ted|GitHub|ypc-faros',
      },
    ]);
  });

  test('flatten nodes V2 with jsonb array', () => {
    const query = `
      query ($from: timestamptz!, $to: timestamptz!) {
        cicd_Artifact (where: {refreshedAt: {_gte: $from, _lt: $to}}) 
          { id tags uid }
      }`;
    const ctx = sut.flattenV2(query, graphSchemaV2);
    for (const [id, type] of ctx.fieldTypes) {
      switch (id) {
        case 'id':
        case 'uid':
          expect(type).toEqual(gql.GraphQLString);
          break;
        case 'tags':
          {
            expect(type.constructor.name).toEqual('GraphQLList');
            const ofType = (type as gql.GraphQLList<gql.GraphQLScalarType>)
              .ofType;
            expect(ofType.name).toEqual('jsonb');
          }
          break;
        default:
          fail(`unexpected field ${id}`);
      }
    }
  });

  test('flatten nodes V2 with list rels', async () => {
    const nodes = [
      {
        incidentUid: 'Q0CYDYKOBEBSPZ',
        impactedApplications: [
          {
            application: {
              applicationName: 'unleash',
            },
          },
        ],
      },
      {
        incidentUid: 'Q0B4CLT0LU9OVT',
        impactedApplications: [
          {
            application: {
              applicationName: 'metabase',
            },
          },
        ],
      },
    ];
    const query = await loadQueryFile('incidents-v2.gql');
    const ctx = sut.flattenV2(query, graphSchemaV2);
    const flattenedNodes = sut.flattenIterable(ctx, nodes);
    expect(ctx.fieldTypes).toEqual(
      new Map([
        ['incident_uid', gql.GraphQLString],
        ['application_name', gql.GraphQLString],
      ])
    );
    expect(await toArray(flattenedNodes)).toIncludeSameMembers([
      {incident_uid: 'Q0CYDYKOBEBSPZ', application_name: 'unleash'},
      {incident_uid: 'Q0B4CLT0LU9OVT', application_name: 'metabase'},
    ]);
  });

  test('flatten nodes V2 with lists within lists', async () => {
    const nodes = [
      {
        number: 37,
        repository: {
          repositoryName: 'airbyte-connectors',
        },
        author: {
          authorUid: 'Roma-Kyrnis',
          identity: [
            {
              identity: {
                authorIdentityName: 'roma',
                employee: [
                  {
                    authorLevel: 'ninja',
                  },
                ],
              },
            },
          ],
        },
      },
      {
        number: 39,
        repository: {
          repositoryName: 'airbyte-connectors',
        },
        author: {
          authorUid: 'tovbinm',
          identity: [
            {
              identity: {
                authorIdentityName: 'matthew',
                employee: [
                  {
                    authorLevel: 'master',
                  },
                ],
              },
            },
          ],
        },
      },
    ];

    const query = await loadQueryFile('pull_request-v2.gql');
    const ctx = sut.flattenV2(query, graphSchemaV2);
    const flattenedNodes = sut.flattenIterable(ctx, nodes);
    expect(await toArray(flattenedNodes)).toMatchSnapshot();
  });

  test('flatten nodes V2 with invalid type', () => {
    expect(() => sut.flattenV2('{foo_Bar{id}}', graphSchemaV2)).toThrow(
      'invalid type \'foo_Bar\''
    );
  });

  test('paginated relay v2 query', async () => {
    const query = await loadQueryFile('commits-v2.gql');
    const expectedQuery = await loadQueryFile('paginated-commits-v2.gql');
    const paginatedQuery = sut.paginatedWithRelayV2(query);
    expect(paginatedQuery.edgesPath).toEqual([
      'vcs_Commit_connection',
      'edges',
    ]);
    expect(paginatedQuery.pageInfoPath).toEqual([
      'vcs_Commit_connection',
      'pageInfo',
    ]);
    expect(paginatedQuery.query).toEqual(expectedQuery);
  });

  test('paginated offset/limit v2 query', async () => {
    const query = await loadQueryFile('commits-v2.gql');
    const paginatedQuery = sut.paginateWithOffsetLimitV2(query);
    const expectedQuery = await loadQueryFile(
      'paginated-commits-offset-limit-v2.gql'
    );
    expect(paginatedQuery.query).toEqual(expectedQuery);
    expect(paginatedQuery.edgesPath).toEqual(['vcs_Commit']);
    expect(paginatedQuery.pageInfoPath).toBeEmpty();
  });

  test('paginated keyset v2 query', async () => {
    const query = await loadQueryFile('incidents-v2.gql');
    const paginatedQuery = sut.paginateWithKeysetV2(query);
    const expectedQuery = await loadQueryFile(
      'paginated-incidents-keyset-v2.gql'
    );
    expect(paginatedQuery.query).toEqual(expectedQuery);
    expect(paginatedQuery.edgesPath).toEqual(['ims_Incident']);
    expect(paginatedQuery.edgeIdPath).toEqual(['_id']);
    expect(paginatedQuery.pageInfoPath).toBeEmpty();
  });

  test('paginated keyset v2 query with existing where clause', async () => {
    const query = await loadQueryFile('commits-v2.gql');
    const paginatedQuery = sut.paginateWithKeysetV2(query);
    const expectedQuery = await loadQueryFile(
      'paginated-commits-keyset-v2.gql'
    );
    expect(paginatedQuery.query).toEqual(expectedQuery);
    expect(paginatedQuery.edgesPath).toEqual(['vcs_Commit']);
    expect(paginatedQuery.edgeIdPath).toEqual(['_id']);
    expect(paginatedQuery.pageInfoPath).toBeEmpty();
  });

  test('build incremental V2', () => {
    const type = graphSchemaV2.getType('cicd_Build') as gql.GraphQLObjectType;
    const query1 = sut.buildIncrementalQueryV2({type});
    expect(query1).toMatchSnapshot();
    const query2 = sut.buildIncrementalQueryV2({
      type: type as gql.GraphQLObjectType,
      avoidCollisions: false,
    });
    expect(query2).toMatchSnapshot();
  });

  test('empty cross merge', async () => {
    expect(await toArray(sut.crossMerge([]))).toBeEmpty();
  });

  test('cross merge with empty iterable', async () => {
    const iters1 = [[{a: 1}, {b: 2}], []];
    expect(await toArray(sut.crossMerge(iters1))).toIncludeSameMembers([
      {a: 1},
      {b: 2},
    ]);

    const iters2 = [[], [{a: 1}, {b: 2}]];
    expect(await toArray(sut.crossMerge(iters2))).toIncludeSameMembers([
      {a: 1},
      {b: 2},
    ]);
  });

  test('single cross merge', async () => {
    const iters = [[{a: 1}, {b: 2}, {c: 3}]];
    expect(await toArray(sut.crossMerge(iters))).toIncludeSameMembers([
      {a: 1},
      {b: 2},
      {c: 3},
    ]);
  });

  test('double cross merge', async () => {
    const iters = [
      [{a: 1}, {b: 2}, {c: 3}],
      [{d: 4}, {e: 5}],
    ];
    expect(await toArray(sut.crossMerge(iters))).toIncludeSameMembers([
      {a: 1, d: 4},
      {a: 1, e: 5},
      {b: 2, d: 4},
      {b: 2, e: 5},
      {c: 3, d: 4},
      {c: 3, e: 5},
    ]);
  });

  test('triple cross merge', async () => {
    const iters = [
      [{a: 1}, {b: 2}, {c: 3}],
      [{d: 4}, {e: 5}],
      [{f: 6}, {g: 7}],
    ];
    expect(await toArray(sut.crossMerge(iters))).toIncludeSameMembers([
      {a: 1, d: 4, f: 6},
      {a: 1, d: 4, g: 7},
      {a: 1, e: 5, f: 6},
      {a: 1, e: 5, g: 7},
      {b: 2, d: 4, f: 6},
      {b: 2, d: 4, g: 7},
      {b: 2, e: 5, f: 6},
      {b: 2, e: 5, g: 7},
      {c: 3, d: 4, f: 6},
      {c: 3, d: 4, g: 7},
      {c: 3, e: 5, f: 6},
      {c: 3, e: 5, g: 7},
    ]);
  });

  test('convert to incremental V2', () => {
    const query = `query MyQuery {
      vcs_PullRequest {
        number
        title
        refreshedAt
      }
    }`;
    expect(sut.toIncrementalV2(query)).toMatchSnapshot();
    const query_without_refreshed_at = `query MyQuery {
      vcs_PullRequest {
        number
        title
      }
    }`;
    expect(sut.toIncrementalV2(query_without_refreshed_at)).toMatchSnapshot();
  });

  test('create incremental queries V2', () => {
    expect(
      sut.createIncrementalQueriesV2({graphSchema: graphSchemaV2})
    ).toMatchSnapshot();
    expect(
      sut.createIncrementalQueriesV2({
        graphSchema: graphSchemaV2,
        avoidCollisions: false,
      })
    ).toMatchSnapshot();
    expect(
      sut.createIncrementalQueriesV2({
        graphSchema: graphSchemaV2,
        avoidCollisions: false,
        scalarsOnly: true,
      })
    ).toMatchSnapshot();
  });

  test('create incremental queries V2 with primary keys/references', () => {
    const primaryKeys = {
      cicd_Build: ['pipeline', 'uid'],
      cicd_Deployment: ['source', 'uid'],
      cicd_Organization: ['source', 'uid'],
      cicd_Pipeline: ['organizationId', 'uid'],
      cicd_Repository: ['organizationId', 'uid'],
    };
    const organizationReferences = {
      organizationId: {
        field: 'organization',
        model: 'cicd_Organization',
        foreignKey: 'organizationId',
      },
      organization: {
        field: 'organization',
        model: 'cicd_Organization',
        foreignKey: 'organizationId',
      },
    };
    const references = {
      cicd_Pipeline: organizationReferences,
      cicd_Repository: organizationReferences,
    };
    expect(
      sut.createIncrementalQueriesV2({
        graphSchema: graphSchemaV2ForPrimaryKeysTest,
        primaryKeys,
        references,
        avoidCollisions: false
      })
    ).toMatchSnapshot();
  });

  test('create incremental queries V2 when missing expected field', () => {
    const primaryKeys = {
      cicd_Build: ['pipeline', 'uid'],
      cicd_Deployment: ['source', 'uid'],
      cicd_Organization: ['source', 'uid'],
      cicd_Pipeline: ['organizationId', 'uid'],
      cicd_Repository: ['organizationId', 'uid'],
    };
    expect(() =>
      sut.createIncrementalQueriesV2({
        graphSchema: graphSchemaV2ForPrimaryKeysTest,
        primaryKeys
      })
    ).toThrowErrorMatchingInlineSnapshot(
      '"expected organizationId to be a field of cicd_Pipeline"'
    );
    const organizationReferences = {
      organizationId: {
        field: 'missing_field',
        model: 'cicd_Organization',
        foreignKey: 'organizationId',
      },
      organization: {
        field: 'missing_field',
        model: 'cicd_Organization',
        foreignKey: 'organizationId',
      },
    };
    const references = {
      cicd_Pipeline: organizationReferences,
      cicd_Repository: organizationReferences,
    };
    expect(() =>
      sut.createIncrementalQueriesV2({
        graphSchema: graphSchemaV2ForPrimaryKeysTest,
        primaryKeys,
        references
      })
    ).toThrowErrorMatchingSnapshot();
  });

  test('excludes foreign key (scalar) identifier from selection', () => {
    const primaryKeys = {
      cicd_Organization: ['source', 'uid'],
      cicd_Pipeline: ['organizationId', 'uid'],
    };
    const organizationReferences = {
      organizationId: {
        field: 'organization',
        model: 'cicd_Organization',
        foreignKey: 'organizationId',
      },
      organization: {
        field: 'organization',
        model: 'cicd_Organization',
        foreignKey: 'organizationId',
      },
    };
    const references = {
      cicd_Pipeline: organizationReferences,
    };
    expect(
      sut.createIncrementalQueriesV2({
        graphSchema: graphSchemaV2ForForeignKeyExclusionTest,
        primaryKeys,
        references,
        avoidCollisions: false
      })
    ).toMatchSnapshot();
  });

  test('path to model V2', () => {
    const query = `query MyQuery {
      cicd_Build {
        number
        name
      }
    }`;
    expect(sut.pathToModelV2(query, graphSchemaV2)).toEqual({
      modelName: 'cicd_Build',
      path: ['cicd_Build'],
    });
  });
});
