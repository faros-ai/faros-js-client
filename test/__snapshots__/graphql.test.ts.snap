// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql build incremental V2 1`] = `
{
  "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Build (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id createdAt endedAt name number origin pipeline { pipelineId: id } refreshedAt startedAt status statusCategory statusDetail uid url } }",
  "name": "cicd_Build",
}
`;

exports[`graphql build incremental V2 2`] = `
{
  "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Build (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id createdAt endedAt name number origin pipeline { id } refreshedAt startedAt status statusCategory statusDetail uid url } }",
  "name": "cicd_Build",
}
`;

exports[`graphql convert to incremental V2 1`] = `
"query incrementalQuery($from: timestamptz!, $to: timestamptz!) {
  vcs_PullRequest(where: {refreshedAt: {_gte: $from, _lt: $to}}) {
    number
    title
    refreshedAt
  }
}"
`;

exports[`graphql convert to incremental V2 2`] = `
"query incrementalQuery($from: timestamptz!, $to: timestamptz!) {
  vcs_PullRequest(where: {refreshedAt: {_gte: $from, _lt: $to}}) {
    number
    title
    refreshedAt
  }
}"
`;

exports[`graphql create incremental queries V2 1`] = `
[
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Build (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id createdAt endedAt name number origin pipeline { pipelineId: id } refreshedAt startedAt status statusCategory statusDetail uid url } }",
    "name": "cicd_Build",
  },
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Pipeline (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id tags } }",
    "name": "cicd_Pipeline",
  },
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Artifact (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id tags uid } }",
    "name": "cicd_Artifact",
  },
]
`;

exports[`graphql create incremental queries V2 2`] = `
[
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Build (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id createdAt endedAt name number origin pipeline { id } refreshedAt startedAt status statusCategory statusDetail uid url } }",
    "name": "cicd_Build",
  },
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Pipeline (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id tags } }",
    "name": "cicd_Pipeline",
  },
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Artifact (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id tags uid } }",
    "name": "cicd_Artifact",
  },
]
`;

exports[`graphql create incremental queries V2 3`] = `
[
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Build (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id createdAt endedAt name number origin refreshedAt startedAt status statusCategory statusDetail uid url } }",
    "name": "cicd_Build",
  },
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Pipeline (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id tags } }",
    "name": "cicd_Pipeline",
  },
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Artifact (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id tags uid } }",
    "name": "cicd_Artifact",
  },
]
`;

exports[`graphql create incremental queries V2 when missing expected field 2`] = `"expected missing_field to be a reference field of cicd_Pipeline (foreign key to cicd_Organization)"`;

exports[`graphql create incremental queries V2 with primary keys/references 1`] = `
[
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Build (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id name pipeline { organization { source uid } uid } uid url } }",
    "name": "cicd_Build",
  },
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Pipeline (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id name organization { source uid } uid url } }",
    "name": "cicd_Pipeline",
  },
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Organization (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id description name origin source uid url } }",
    "name": "cicd_Organization",
  },
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Repository (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id description name organization { source uid } origin uid url } }",
    "name": "cicd_Repository",
  },
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Deployment (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id source statusCategory statusDetail uid url } }",
    "name": "cicd_Deployment",
  },
]
`;

exports[`graphql excludes foreign key (scalar) identifier from selection 1`] = `
[
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Pipeline (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id name organization { source uid } uid url } }",
    "name": "cicd_Pipeline",
  },
  {
    "gql": "query ($from: timestamptz!, $to: timestamptz!) { cicd_Organization (where: {refreshedAt: {_gte: $from, _lt: $to}}) { id description name origin source uid url } }",
    "name": "cicd_Organization",
  },
]
`;

exports[`graphql flatten nodes V2 with lists within lists 1`] = `
[
  {
    "author_identity_name": "roma",
    "author_level": "ninja",
    "author_uid": "Roma-Kyrnis",
    "number": 37,
    "repository_name": "airbyte-connectors",
  },
  {
    "author_identity_name": "matthew",
    "author_level": "master",
    "author_uid": "tovbinm",
    "number": 39,
    "repository_name": "airbyte-connectors",
  },
]
`;
